File: crates/kuben-platform/src/health.rs
//! Subsystem health registry backing `/livez`, `/readyz` and
//! `/healthz/details`.

use std::{
    sync::{
        Arc,
        atomic::{AtomicBool, AtomicI64, Ordering},
    },
    time::{SystemTime, UNIX_EPOCH},
};

use dashmap::DashMap;
use serde::Serialize;

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum State {
    Ok,
    Degraded,
    Starting,
    /// Healthy but deliberately idle, e.g. waiting for the controller lease.
    Standby,
}

#[derive(Clone, Debug, Serialize)]
pub struct Subsystem {
    pub state: State,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    pub updated_at_ms: i64,
}

#[derive(Clone, Debug, Default)]
pub struct Health {
    inner: Arc<Inner>,
}

#[derive(Debug)]
struct Inner {
    ready: AtomicBool,
    /// Last watchdog heartbeat (unix ms). `/livez` fails if this goes stale.
    heartbeat: AtomicI64,
    subsystems: DashMap<&'static str, Subsystem>,
}

impl Default for Inner {
    fn default() -> Self {
        Self {
            ready: AtomicBool::new(false),
            heartbeat: AtomicI64::new(now_ms()),
            subsystems: DashMap::new(),
        }
    }
}

fn now_ms() -> i64 {
    i64::try_from(
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_or(0, |d| d.as_millis()),
    )
    .unwrap_or(i64::MAX)
}

impl Health {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    pub fn set_ready(&self, ready: bool) {
        self.inner.ready.store(ready, Ordering::SeqCst);
    }

    #[must_use]
    pub fn is_ready(&self) -> bool {
        self.inner.ready.load(Ordering::SeqCst)
    }

    /// Called by the watchdog task every second.
    pub fn heartbeat(&self) {
        self.inner.heartbeat.store(now_ms(), Ordering::Relaxed);
    }

    /// Liveness: the runtime has heartbeated within `max_age_ms`.
    #[must_use]
    pub fn is_live(&self, max_age_ms: i64) -> bool {
        now_ms() - self.inner.heartbeat.load(Ordering::Relaxed) < max_age_ms
    }

    pub fn starting(&self, name: &'static str) {
        self.set(name, State::Starting, None);
    }

    pub fn ok(&self, name: &'static str) {
        self.set(name, State::Ok, None);
    }

    pub fn standby(&self, name: &'static str) {
        self.set(name, State::Standby, None);
    }

    pub fn degraded(&self, name: &'static str, error: &str) {
        self.set(name, State::Degraded, Some(error.to_owned()));
    }

    fn set(&self, name: &'static str, state: State, last_error: Option<String>) {
        self.inner.subsystems.insert(
            name,
            Subsystem {
                state,
                last_error,
                updated_at_ms: now_ms(),
            },
        );
    }

    /// Snapshot of every subsystem, sorted by name.
    #[must_use]
    pub fn details(&self) -> Vec<(&'static str, Subsystem)> {
        let mut v: Vec<_> = self
            .inner
            .subsystems
            .iter()
            .map(|e| (*e.key(), e.value().clone()))
            .collect();
        v.sort_by_key(|(k, _)| *k);
        v
    }

    #[must_use]
    pub fn any_degraded(&self) -> bool {
        self.inner
            .subsystems
            .iter()
            .any(|e| e.value().state == State::Degraded)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

File: go/hub/internal/platform/health/health.go
// Package health is the subsystem health registry behind /livez, /readyz
// and /api/v1/healthz/details (crates/kuben-platform/src/health.rs).
package health

import (
	"maps"
	"slices"
	"sync"
	"sync/atomic"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/clock"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/metrics"
)

// State is how a subsystem is doing.
type State string

// The states.
const (
	Ok       State = "ok"
	Degraded State = "degraded"
	Starting State = "starting"
	// Standby is healthy but deliberately idle, e.g. waiting for the
	// controller lease.
	Standby State = "standby"
)

// Subsystem is the reported health of one part of the server.
type Subsystem struct {
	State       State           `json:"state"`
	LastError   opt.Val[string] `json:"last_error,omitzero"`
	UpdatedAtMs int64           `json:"updated_at_ms"`
}

// Health is shared by every part of the server; the zero Health is not
// usable, make one with New.
type Health struct {
	clock     clock.Clock
	ready     atomic.Bool
	heartbeat atomic.Int64 // last watchdog beat (unix ms); /livez fails when stale

	mu         sync.Mutex // guards subsystems
	subsystems map[string]Subsystem

	metrics *metrics.Metrics // the process's metrics, beside its health
}

// New is a registry that is not ready and has just beaten.
func New(c clock.Clock) *Health {
	h := &Health{clock: c, subsystems: map[string]Subsystem{}, metrics: metrics.New()}
	h.heartbeat.Store(c.NowMs())
	return h
}

// Metrics are the process's Prometheus metrics; every part of the server
// that reports health reports its counters here too. A nil Health has none
// (nil metrics record nothing).
func (h *Health) Metrics() *metrics.Metrics {
	if h == nil {
		return nil
	}
	return h.metrics
}

// SetReady sets readiness.
func (h *Health) SetReady(ready bool) { h.ready.Store(ready) }

// IsReady reports readiness.
func (h *Health) IsReady() bool { return h.ready.Load() }

// Heartbeat is called by the watchdog every second.
func (h *Health) Heartbeat() { h.heartbeat.Store(h.clock.NowMs()) }

// IsLive reports whether the watchdog beat within maxAgeMs.
func (h *Health) IsLive(maxAgeMs int64) bool {
	return h.clock.NowMs()-h.heartbeat.Load() < maxAgeMs
}

// Starting marks name as starting.
func (h *Health) Starting(name string) { h.set(name, Starting, opt.None[string]()) }

// OK marks name as healthy.
func (h *Health) OK(name string) { h.set(name, Ok, opt.None[string]()) }

// Standby marks name as deliberately idle.
func (h *Health) Standby(name string) { h.set(name, Standby, opt.None[string]()) }

// Degrade marks name as degraded by err.
func (h *Health) Degrade(name, err string) { h.set(name, Degraded, opt.Some(err)) }

func (h *Health) set(name string, s State, lastError opt.Val[string]) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.subsystems[name] = Subsystem{State: s, LastError: lastError, UpdatedAtMs: h.clock.NowMs()}
}

// Named is a subsystem with its name.
type Named struct {
	Name string
	Subsystem
}

// Details is every subsystem, sorted by name.
func (h *Health) Details() []Named {
	h.mu.Lock()
	defer h.mu.Unlock()
	names := slices.Sorted(maps.Keys(h.subsystems))
	out := make([]Named, 0, len(names))
	for _, n := range names {
		out = append(out, Named{Name: n, Subsystem: h.subsystems[n]})
	}
	return out
}

// AnyDegraded reports whether a subsystem is degraded.
func (h *Health) AnyDegraded() bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	for _, s := range h.subsystems {
		if s.State == Degraded {
			return true
		}
	}
	return false
}
